out_writeapi.go

```diff
@@ -580,14 +609,17 @@ func FLBPluginFlushCtx(ctx, data unsafe.Pointer, length C.int, tag *C.char) int
 		}
 
 		if (currsize + len(buf)) >= config.maxChunkSize {
+			streamIndex := getStreamIndex()
```

Comment:
Might we select a different stream across chunks? Alternatively we could select the stream up-front and use it for both chunking and non-chunking cases. What are the tradeoffs?

Reply:
I believe we can select different streams across chunks in this current implementation. I believe this would be especially useful in cases where there is a lot of input data that is being chunked into a lot of smaller pieces. If we picked a stream up front and used it for the chunking and non-chunking portions, there is a chance that one stream might still have the same issue of back pressure due to AppendRows being called in a for loop. By selecting a different stream across chunks, we can take advantage and lower the amount of requests in a single stream.

Reply:
A trade off would be this makes the ability to guarantee Exactly Once a bit trickier, but not impossible.

Reply:
In practice, our default chunk size is 9MB and Fluentbit sends us data of at most 2MB in a single call to Flush. Hence chunking will typically never happen. The code would be somewhat simpler if we identified the target stream up-front, because then we need to call this logic in only 1 place rather than in 2 places.

Exactly-once is managed per stream so will work with both approaches.

Again, either approach will work.